core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaStoreMessageHandler.java

```diff
@@ -85,6 +93,29 @@ private void handleSchemaUpdate(SchemaKey schemaKey, SchemaValue schemaObj) {
       }
       schemaIdAndSubjects.addSubjectAndVersion(schemaKey.getSubject(), schemaKey.getVersion());
       schemaRegistry.schemaHashToGuid.put(md5, schemaIdAndSubjects);
+
+      if (schemaObj.isDeleted()) {
+        handleDeleteSchemaKey(schemaKey, schemaObj);
+      } else {
+        List<SchemaKey> schemaKeys = schemaRegistry.guidToDeletedSchemaKeys
+            .getOrDefault(schemaObj.getId(), Collections.emptyList());
+        schemaKeys.stream().filter(v -> v.getSubject().equals(schemaObj.getSubject()))
+            .forEach(k -> tombstoneExecutor.execute(() -> tombstoneSchemaKey(k)));
```

Comment:
If I understand this logic, then when the schema is being deleted we will record the GUID for that schema and include the `SchemaKey` in the list of deleted schemas.

Then, when the schema is _not_ being deleted, we want to check whether a prior schema existed with a `SchemaKey` that has the same subject. If so, then we want to record in the Kafka topic a tombstone for this schema before this new schema is stored.

But this code writes the tombstone _asynchronously_, so how do we guarantee that the tombstone is written before the new schema?

Reply:
BTW, comments would be helpful for this whole block. For example, why are we only writing the tombstone when a new schema is registered for a previously deleted schema? We don't want to record the tombstone every time a schema is deleted? The comments should address this, since these assumptions/constraints are not defined elsewhere.

Reply:
Also, why not use
```
.forEach(this::tombstoneSchemaKey);
```
and then make `tombstoneSchemaKey` call the executor internally?

Reply:
It's not necessary for the tombstone to be written before new schema is stored. That's the reason it can be asynchronous. Technically, the schema version is already deleted but from a storage perspective we do a soft delete so that id can still be accessed by old consumers. Tombstone is to optimize the actual physical storage.

Reply:
Any reason why we don't just do the writes to Kafka asynchronously when the delete request comes in? I understand there is no impact if we lose tombstone writes (since `guidToDeletedSchemaKeys` can be lost during SR restarts), but IMHO it's easier to reason about the overall code flow.